Authorize more resources

jgreben opened the pull request:
This PR should make it so that direct URL access to open routes are disallowed.

jgreben committed:
authorize resources for various controllers to prevent direct url access

shelleydoljack commented:
I tried creating a shelf selection report and kept getting the error "Loc array can't be empty" even though I entered a LC range. Here's the log file for the report:
>10:48:10  Begin shelf_sel.pl
Full command line
/s/SUL/Bin/ShelfSel/shelf_sel.pl 'sdoljack@stanford.edu' 'sdoljack' 'A-B' 'GREEN' '' '0' 'All Formats' 'All Item Types' 'All Item Category1s' '' '' '' '' '' '2' '0' '0' '0' '0' '0' '0' '2' '' 'https://libsys-webforms-dev.stanford.edu'
10:48:11  Finished reading LC selcallnum source (1-letter args) A-B
167907 lines selected
10:48:12  shelf_sel.pl fatal error
SIGPIPE writing to selitem
Broken pipe

shelleydoljack commented:
Clicking on "Batch details" for the SAL3 batch requests created an error:
https://libsys-webforms-dev.stanford.edu/sal3_batch_requests_batches/52
`undefined local variable or method `sal3_batch_request_bc_path' `

shelleydoljack commented:
I test submitted a SAL3 batch request and even though the delivery start and deliver by date fields were filled in with default values `FEB/20/21` and `AUG/20/21` the form alerted me to fill in these dates. I used the datepicker and the date format was `MM/DD/YYYY` and the form was able to be submitted.

shelleydoljack reviewed (changes requested):
Thanks for figuring this out! I removed my user from the authorized_user table and checked all the routes that I identified in https://github.com/sul-dlss/LibSysTasks/issues/566 . I got the expected flash message of not being authorized to view the page. 

There were some issues I came across that are separately commented that need addressing. Also, I wonder if we want to create an ability_spec, somewhat like I did in my WIP branch https://github.com/sul-dlss/libsys-webforms/blob/fix-abilities/spec/models/ability_spec.rb .

jgreben commented:
> I tried creating a shelf selection report and kept getting the error "Loc array can't be empty" even though I entered a LC range. Here's the log file for the report:

> > /s/SUL/Bin/ShelfSel/shelf_sel.pl '[sdoljack@stanford.edu](mailto:sdoljack@stanford.edu)' 'sdoljack' 'A-B' 'GREEN' '' '0' 'All Formats' 'All Item Types' 'All Item Category1s' '' '' '' '' '' '2' '0' '0' '0' '0' '0' '0' '2' '' 'https://libsys-webforms-dev.stanford.edu'

`loc_array` is the home location field on the report and it cannot be empty. Looks like that would come after 'GREEN' in the params above, but it is an empty string.

shelleydoljack commented:
> `loc_array` is the home location field on the report and it cannot be empty. Looks like that would come after 'GREEN' in the params above, but it is an empty string.

I must have been tired while testing this, thinking "loc" stood for Library of Congress. The report works when I select a home location. It seems like adding more meaningful error messages would be helpful, but in another PR probably. Also, the management report forms alignment needs fixing (also good in another PR).

jgreben commented:
@shelleydoljack I think I fixed all of the problems you noted. The date problem had to do with the format that the HTML5 datepicker expects. The `sal3_batch_requests_bcs_path` had to do with the fact that I changed the controller name to be plural (requests vs request) for consistency sake, and I didn't find all of the places where I needed to change it. I made a note in the ticket about more integration tests to specifically add more tests that would have caught this problem.

jgreben commented:
> There were some issues I came across that are separately commented that need addressing. Also, I wonder if we want to create an ability_spec, somewhat like I did in my WIP branch https://github.com/sul-dlss/libsys-webforms/blob/fix-abilities/spec/models/ability_spec.rb .

Do you want to push up a PR with that `ability_spec` ?

shelleydoljack commented:
> Do you want to push up a PR with that `ability_spec` ?

I'll need to update it a bit, I think, but yes, I'll do that.

shelleydoljack commented:
Looks good. One thing that maybe @dlrueda can clarify regarding the batch record updates menu. In her comment https://github.com/sul-dlss/LibSysTasks/issues/566#issuecomment-582016876 Darsi says
>For the Batch Record Updates menu, there are some parts of it that should be open to viewing by anyone that can webauth: show queue, show batch errors, show MHLD errors
So, nothing on the "Upload a batch" menu, but everything else, yes.

After removing my user from the authorized user table, I am able to access direct links to showing completed and incomplete batches, as well as batch and mhld errors (such as https://libsys-webforms-dev.stanford.edu/batch_record_updates/errors_for_mhld?utf8=✓&batch_number=11&commit=Go) but I cannot view anything in the batch record updates menu. Since users get email reports with direct links to the errors, do they need to view the batch record update menu (all sections except "Upload a batch" section) or do the direct links in emails suffice?

dlrueda commented:
Shelley asks:

> Since users get email reports with direct links to the errors, do they need to view the batch record update menu (all sections except "Upload a batch" section) or do the direct links in emails suffice?


I mean, I suppose maybe not, since they usually have the link in an email. But really it _should_ be available to them if they go directly to the web page.  Then they'd just remember the batchID and put it in.

Is this technically difficult?

shelleydoljack commented on app/models/ability.rb:
```diff
@@ -40,14 +41,29 @@ def assign_staff_read_permission(current_user)
   def assign_staff_manage_permission(current_user)
     can :manage, AccessionNumberUpdate if /A|Y/.match?(current_user.accession_number)
     can :manage, EdiInvoice if /A|Y/.match?(current_user.edi_inv_manage)
     can :manage, EdiErrorReport if /A|Y/.match?(current_user.edi_inv_manage)
```
If I have edi_inv_view = "Y" I cannot view the EDI error reports. Is that okay? Or is it expected that I would have edi_inv_manage = "Y" in order to view errors?

dlrueda replied:
If you have edi_inv_view = 'Y' you SHOULD be able to view the EDI error reports.  (Otherwise there's not much to do as a "view" user.)

jgreben replied:
I added some changes to allow staff with edi_inv_view to read the error reports.

shelleydoljack replied:
I can now view errors with edi_inv_view='Y'. I can also edit by going directly to `edi_inv_line/837865/edit` even though I have edi_inv_manage=null. I believe I can edit because I have mgt_rpts='A'.

jgreben replied:
After making the change to `Ability` to move the Edi permissions out of the mgt_rpts section, I tried it with the `atest` user (with edi_inv_manage: "") and cancan prevents the edit.

shelleydoljack replied:
It works now. Thanks.

shelleydoljack reviewed (commented):
Another question about edi_inv_view and edi_inv_manage. With edi_inv_view = "Y" I cannot view EDI errors, is that expected?

jgreben committed:
Add ability for any user to view batch record updates

shelleydoljack commented on app/models/ability.rb:
```diff
@@ -38,16 +42,32 @@ def assign_staff_create_permission(current_user)
+
+    can :manage, CirculationStatisticsReport
+    can :manage, EdiInvLine
+    can :manage, EdiLin
```
The EdiInvLine and EdiLin should probably be separated since authorized_user.edi_inv_manage should determine manage permissions for these and not authorized_user.mgt_rpts.

shelleydoljack commented on app/models/ability.rb:
```diff
@@ -18,6 +19,9 @@ def initialize(current_user)
 
   def assign_basic_permission
     can :manage, Ckey2bibframe
+    can :read, UniUpdatesErrors
+    can :read, UniUpdMhldError
+    can :read, BatchRecordUpdate
```
This now allows me to view the Batch Record Updates menu on the main page and the show actions on the batch record updates menu page when I remove my user from authorized_user. Thanks!

jgreben replied:
Great!

jgreben committed:
Seperate out EDI for manage permissions
